target/ppc/mem_helper.c

```diff
@@ -369,50 +369,37 @@ target_ulong helper_lscbx(CPUPPCState *env, target_ulong addr, uint32_t reg,
 }
 
 #ifdef TARGET_PPC64
-uint64_t helper_lq_le_parallel(CPUPPCState *env, target_ulong addr,
-                               uint32_t opidx)
+uint64_t helper_lq_parallel(CPUPPCState *env, target_ulong addr, uint32_t opidx)
 {
     Int128 ret;
 
     /* We will have raised EXCP_ATOMIC from the translator.  */
     assert(HAVE_ATOMIC128);
-    ret = cpu_atomic_ldo_le_mmu(env, addr, opidx, GETPC());
-    env->retxh = int128_gethi(ret);
-    return int128_getlo(ret);
-}
-
-uint64_t helper_lq_be_parallel(CPUPPCState *env, target_ulong addr,
-                               uint32_t opidx)
-{
-    Int128 ret;
-
-    /* We will have raised EXCP_ATOMIC from the translator.  */
-    assert(HAVE_ATOMIC128);
-    ret = cpu_atomic_ldo_be_mmu(env, addr, opidx, GETPC());
-    env->retxh = int128_gethi(ret);
-    return int128_getlo(ret);
-}
-
-void helper_stq_le_parallel(CPUPPCState *env, target_ulong addr,
-                            uint64_t lo, uint64_t hi, uint32_t opidx)
-{
-    Int128 val;
-
-    /* We will have raised EXCP_ATOMIC from the translator.  */
-    assert(HAVE_ATOMIC128);
-    val = int128_make128(lo, hi);
-    cpu_atomic_sto_le_mmu(env, addr, val, opidx, GETPC());
+    if (msr_le) {
+        ret = cpu_atomic_ldo_le_mmu(env, addr, opidx, GETPC());
+        env->retxh = int128_gethi(ret);
+        return int128_getlo(ret);
+    } else {
+        ret = cpu_atomic_ldo_be_mmu(env, addr, opidx, GETPC());
+        env->retxh = int128_getlo(ret);
+        return int128_gethi(ret);
+    }
 }
 
-void helper_stq_be_parallel(CPUPPCState *env, target_ulong addr,
-                            uint64_t lo, uint64_t hi, uint32_t opidx)
+void helper_stq_parallel(CPUPPCState *env, target_ulong addr, uint64_t lo,
+                         uint64_t hi, uint32_t opidx)
 {
     Int128 val;
 
     /* We will have raised EXCP_ATOMIC from the translator.  */
     assert(HAVE_ATOMIC128);
-    val = int128_make128(lo, hi);
-    cpu_atomic_sto_be_mmu(env, addr, val, opidx, GETPC());
+    if (msr_le) {
+        val = int128_make128(lo, hi);
+        cpu_atomic_sto_le_mmu(env, addr, val, opidx, GETPC());
+    } else {
+        val = int128_make128(hi, lo);
```

Comment:
Not exactly sure how to solve this, but I find the use of the names lo/hi here (and, similarly, tgt_a/tgt_b in do_ldst_quad) makes the code hard do read, and should be improved.

They don't refer to the low 64-bits (least significant bits) or high 64-bits (most significant bits) of the data, but rather to the data that comes from/goes to the lower address in memory (in the case of [p]{lq,stq}, the data from/to EA and EA+8, respectively).

Their names should be clearer about their true meaning.
(also, 'tgt_' isn't a good prefix for do_ldst_quad, because although it refers to target register for loads, it refers to source registers for stores)

The hard part is coming up with good names, though. Here's an idea, but the names are kind of long. :)
lo/tgt_a could be renamed to val_in_low_mem_addr/low_mem_addr_gpr
hi/tgt_b could be renamed to val_in_high_mem_addr/high_mem_addr_gpr

Or maybe lo_addr/hi_addr? Not completely clear, though.

Reply:
Yeah, these names are terrible. I used a/b in the name because a gave up on finding something better :)
Maybe we should just call it `hi`/`lo` and put a comment telling that it refers to the address...